lib/switchs.nix

```diff
@@ -0,0 +1,75 @@
+{ lib }:
+with lib; {
+  /* Emulate a "switch - case" construct,
+   instead of relying on `if then else if ...` */
+
+  /* Usage:
+       ```nix
+       switch-if [
+         if-clause-1
+         ..
+         if-clause-k
+       ] default-out
+       ```
+       where a if-clause has the form `{ cond = b; out = r; }`
+       the first branch such as `b` is true
+
+     Example:
+       ```nix
+       let alphanum = n: switch-if [
+           { cond = n == 0; out = "zero"; }
+           { cond = n == 1; out = "one";  }
+         ] "I do not count beyond one"; in
+       [ (alphanum 0) (alphanum 1) (alphanum 42) ]
+       ```
+      => `[ "zero" "one" "I do not count beyond one" ]`
+
+  */
+  switch-if = c: d: (findFirst (getAttr "cond") {} c).out or d;
+
+  /* Usage:
+       ```nix
+       switch x [
+         simple-clause-1
+         ..
+         simple-clause-k
+       ] default-out
+       ```
+       where a simple-clause has the form `{ case = p; out = r; }`
+       the first branch such as `p x` is true
+       or
+       ```nix
+       switch [ x1 .. xn ] [
+         complex-clause-1
+         ..
+         complex-clause-k
+       ] default-out
+       ```
+       where a complex-clause is either a simple-clause
+       or has the form { cases = [ p1 .. pn ]; out = r; }
+       in which case the first branch such as all `pi x` are true
+
+       if the variables p are not functions,
+       they are converted to a `equal p`
+       if `out` is missing then `default-out` is taken
```

Comment:
Is there a use case for a `complex-clause` without an `out`?

Reply:
I think I have one somewhere, but anyway, this makes the code shorter!

Reply:
Does it? I thought removing this would make this code:

```nix
    switch-if (map (cl: { cond = combine cl var;
                          out = cl.out or default; }) clauses) default;
```

become:

```nix
    switch-if (map (cl: cl // { cond = combine cl var; }) clauses) default;
```

But if you actually use this, never mind.